test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/ModelBinderFactoryTest.cs

```diff
@@ -235,6 +236,244 @@ public void CreateBinder_Caches_WhenTokenIsNotNull()
             Assert.Same(result1, result2);
         }
 
+        [Fact]
+        public void CreateBinder_Caches_InteriorNodes()
+        {
+            // Arrange
+            var metadataProvider = new TestModelMetadataProvider();
+
+            var options = new TestOptionsManager<MvcOptions>();
+
+            IModelBinder inner = null;
+
+            var widgetProvider = new TestModelBinderProvider(c =>
+            {
+                if (c.Metadata.ModelType == typeof(Widget))
+                {
+                    var binder = c.CreateBinder(c.Metadata.Properties[nameof(Widget.Id)]);
+                    if (inner == null)
+                    {
+                        inner = binder;
+                    }
+                    else
+                    {
+                        Assert.Same(inner, binder);
+                    }
+
+                    return Mock.Of<IModelBinder>();
+                }
+
+                return null;
+            });
+
+            var widgetIdProvider = new TestModelBinderProvider(c =>
+            {
+                Assert.Equal(typeof(WidgetId), c.Metadata.ModelType);
+                return Mock.Of<IModelBinder>();
+            });
+
+            options.Value.ModelBinderProviders.Add(widgetProvider);
+            options.Value.ModelBinderProviders.Add(widgetIdProvider);
+
+            var factory = new ModelBinderFactory(metadataProvider, options);
+
+            var context = new ModelBinderFactoryContext()
+            {
+                Metadata = metadataProvider.GetMetadataForType(typeof(Widget)),
+                CacheToken = null, // We want the provider to run twice.
+            };
+
+            // Act
+            var result1 = factory.CreateBinder(context);
+            var result2 = factory.CreateBinder(context);
+
+            // Assert
+            Assert.NotSame(result1, result2);
+
+            Assert.Equal(2, widgetProvider.SuccessCount);
+            Assert.Equal(1, widgetIdProvider.SuccessCount);
+        }
+
+        [Fact]
+        public void CreateBinder_Caches_InteriorNodes_WhenInteriorNodeReturnsNull()
+        {
+            // Arrange
+            var metadataProvider = new TestModelMetadataProvider();
+
+            var options = new TestOptionsManager<MvcOptions>();
+
+            IModelBinder inner = null;
+
+            var widgetProvider = new TestModelBinderProvider(c =>
+            {
+                if (c.Metadata.ModelType == typeof(Widget))
+                {
+                    var binder = c.CreateBinder(c.Metadata.Properties[nameof(Widget.Id)]);
+                    Assert.IsType<NoOpBinder>(binder);
+                    if (inner == null)
+                    {
+                        inner = binder;
+                    }
+                    else
+                    {
+                        Assert.Same(inner, binder);
+                    }
```

Comment:
This won't be an instance of `NoOpBinder`, it's the mock returned by the provider for `Widget.Id`